nuc_morph_analysis/lib/preprocessing/twoD_zMIP_area/watershed_workflow.py

```diff
@@ -0,0 +1,108 @@
+     output_directory : str
+        the output directory to save the pseudo cell images
+    parallel : bool
+        whether to run the analysis in parallel or not
+        NOTE: currently does not work when reading S3 files
+    save_df : bool
+        whether to save the pseudo cell boundaries as a csv, default is False
+    testing : bool
+        whether to run the function in testing mode, default is False
+        in testing mode only 20 frames are run
+    """
+    # load the segmentation iamge
+    reader = load_data.get_dataset_segmentation_file_reader(colony)
+    args = [(timepoint, colony, resolution_level) for timepoint in range(reader.dims.T)]
+    del reader # close the reader so it doesn't have pickling issues in multiprocessing
+
+    nworkers = cpu_count()
+    if testing:
+        args = args[0:nworkers*2]
+
+    if parallel==False:
+        dflist = []
+        for ai in tqdm(range(len(args)), desc="Processing timepoints"):
+            df_2d = process_timepoint(args[ai])
+            dflist.append(df_2d)
+    else:
+        step = nworkers*2
+        arg_set = [args[ai:ai+step] for ai in range(0,len(args),step)]
+        print("number of sets to process:",len(arg_set))
+        print("step size:",step)
+        dflist_list = []
+        # for arg_subset in tqdm(arg_set, desc="Processing timepoints"):
+        ai = 0
+        while ai < len(arg_set):
+            # code is set up this way to retry if encountrering aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
+            try:
+                arg_subset = arg_set[ai]
+                with get_context('spawn').Pool(nworkers) as p:
+                    dflist = list(tqdm(p.imap_unordered(process_timepoint, arg_subset), initial= ai*step, total=len(args), desc="Processing timepoints",leave=False))
+                    dflist_list.extend(dflist)
+                ai += 1
+            
+            # catch exceptions and try again (kill if keyboard interrupt)
+            except Exception as e:
+                if isinstance(e, KeyboardInterrupt):
+                    print("Keyboard interrupt detected, killing pool")
+                    p.terminate()
+                    p.join()
+                    break
+                else:
+                    p.terminate()
+                    p.join()
+                print("Exception detected, trying again")
+
+    # concatenate the dataframe
+    df = pd.concat(dflist_list)
+
+    if save_df:
+        # save the dataframe as a parquet
+        df.to_parquet(output_directory / f"{colony}_pseudo_cell_boundaries.parquet")
+        print('saved pseudo cell boundaries to:', output_directory / f"{colony}_pseudo_cell_boundaries.parquet")
+
+    return df
+
+if __name__ == "__main__":
+    for colony in load_data.get_dataset_names(dataset='all_baseline'):
+        print(colony)
+        output_directory = Path(__file__).parents[6] / "local_storage" / "pseudo_cell_boundaries"
```

Comment:
the saving is NOT default behavior and is only for troubleshooting. So I think it's fine to go to a non-standard place since it is a non-standard usage.